### ISSUE TITLE
[feature request] Add total episodes parameter to model learn method

### ISSUE THREAD

**Issue by xicocaio** [DLR-RM/stable-baselines3#140]
Hi,

**TL;DR:** I would like the option to pass either a `total_episodes` parameter or a `total_timesteps` to the `model.learn()` method.

Now, for my reasoning. Currently, we can only define the `total_timesteps` when training an agent, as follows.

```
model = A2C('MlpLstmPolicy', env, verbose=1, policy_kwargs=policy_kwargs)
model.learn(total_timesteps=1000)
```

However, for some scenarios (e.g., stock trading), it is quite common to have a fixed number of timesteps per episode, given by the available time-series data points. Also, it can be quite valuable to scan all data points an equal amount of time thoroughly and to determine the number of passes, which is defined by the number of episodes.

Thus, to train for a given number of episodes for a fixed number of timesteps, I have to get the `total_timesteps` value, before passing it to method `model.learn()` as follows:

```
desired_total_episodes = 100
n_points = train_df.shape[0]) # get the number of data points
total_timesteps = desired_total_episodes * n_points
```

Even so, [this answer on StackOverflow](https://stackoverflow.com/a/56766128/2464671) says that

> Where the episode length is known, set it to the desired number of episode you would like to train. However, it might be less because **the agent might not (probably wont) reach max steps every time**.

I must admit I do not know how accurate this answer, but this worries me that my model may not scan all the data equally.

Another option, as discussed in this [issue from previous stable baseline repo](https://github.com/hill-a/stable-baselines/issues/571), is to use a callback function. Still, for this callback approach, I would have to pass a `total_timesteps` variable that is high enough so that I can have the desired number of episodes. Hence, this callback approach seems like an out of the way workaround.

In conclusion, I believe that including the option to pass a `total_episodes` could be a simple and effective approach that would broaden the number of use cases attended by this project.

Thank you for your attention!

**Comment by Miffyli** [comment]
I gave this some more, fresh thought:

- If episode lengths are known and fixed, then the suggested method should work. I still do not understand the cited Stackoverflow message and what they'd be referring to. If episodes always last for N steps, and you want to play M games, and you set `total_timesteps = N * M`, then the agent will be trained exactly for M games. No episode should cut short and once training has ended each environment should have played equal amount of games.
- If episode lengths are unknown, the suggest callback method works and computationally takes the same amount of time: the callback's `step` method is always called, even if you do not provide a callback, and beyond that point the program logic would be the same. `total_timesteps` can be set to infinity or a very high number so it should not interfere, **however** any schedule based on training progress (e.g. learning rate schedules, DQN epsilon-greedy exploration) would not work correctly as they are based on "how far are we in the training progression". This could be fixed to work with "how many episodes of wanted amount of episodes have been played", but that would require quite a bit of new logic.

In total I am not sure if this would be an useful change for many, especially given the potential confusion with the schedulers.

**Comment by xicocaio** [comment]
> No episode should cut short and once training has ended each environment should have played equal amount of games.

That's great, I was hoping for this clarification and assurance of the desired behavior.

> `total_timesteps` can be set to infinity or a very high number so it should not interfere, **however** any schedule based on training progress (e.g. learning rate schedules, DQN epsilon-greedy exploration) would not work correctly as they are based on "how far are we in the training progression". This could be fixed to work with "how many episodes of wanted amount of episodes have been played", but that would require quite a bit of new logic.

I understand this issue, and that's why I realize that the ideal solution would assume both:  

1. Calculating `total_timesteps = N * M`: to give an upper bound of timesteps and, thus, avoid breaking the mentioned algorithms
2. Counting max episodes through callback:  to avoid running more episodes than desired for the case one introduces more conditions for ending the episode (e.g., the RL agent going broke by reaching a minimum amount of net worth).

To wrap up, I understand that your suggested workarounds are not particularly hard to implement and any user of your project should be able to apply them. However, looking through the optics of a practitioner for whom adopting the number of episodes might be more practical than using the timesteps parameter (e.g., stock trading), wouldn't it be more straightforward to

- abstract these workarounds from the user and add an _optional_ `max_episodes` parameter while keeping a _required_ `total_timesteps`?
- or even just add a note to the documentation regarding these cases and workarounds?

**Comment by Miffyli** [comment]
The 1) part should already fix this, unless you have additional conditions to terminate the training as mentioned in 2). I think this could be documented or even made as a pre-made callback, but adding the `max_episodes` parameter would not be as tidy for the reasons mentioned above (confusion with `total_timesteps` and schedules, more code that keeps track of episodes). A pre-made callback would be as easy to use (one more line + one parameter to `learn`). If you make such a callback we could include it via a PR.

**Comment by xicocaio** [comment]
Ok, that sounds good.

But before I begin, I am thinking about using the `StopTrainingOnRewardThreshold` callback as my reference, do you think that's good or do you suggest any other approach?

**Comment by Miffyli** [comment]
That should work as a good starting point, yes :). The behaviour is very similar, just need to track number of episodes played rather than rewards.

**Comment by xicocaio** [comment]
Good.

Also, In the `StopTrainingOnRewardThreshold` callback documentation it says that "It must be used with the `EvalCallback`."

I believe I could probably drop this requirement of also using the `EvalCallback` as I only want to count episodes and do not need to keep track of performance metrics, right?

**Comment by Miffyli** [comment]
Yes, you do not need the `EvalCallBack` for this one, as it was only used to see how well agent was doing.

**Comment by xicocaio** [comment]
Ok, thank you! =)

**Comment by xicocaio** [comment]
Ok, thank you! =)

I started working on the present issue today, and I need some clarification.

Regarding keeping count of the number of episodes, initially, I thought about accessing the return from `env.step()`. However, as I understand this might not yet be possible as reported in #137 .

As a workaround, I am experimenting with using the model's property `self.model._episode_num`. Still, I do not feel comfortable accessing a variable supposed to be private. Is this the best approach? What would you suggest?

**Comment by Miffyli** [comment]
Callback's are meant to be bit dirty so accessing a private variable (with caution) is alright in this case, however I am not certain if that variable is available with all agents. I think the safest bet is to work based on #137 and assume the local variables are available. That should across agents.

**Comment by xicocaio** [comment]
I just verified that indeed for the `A2C` and `PPO` models the `self.model._episode_num` is not being incremented.

#137 is labeled as a question which points to #115. However, #115 is a PR from a forked repo and has not been accepted yet. So just clarifying, your suggestion is that I use the unmerged forked code from this PR?

Regarding the merge note, I don't think I fully understand the implications of that. You mean in the sense that I will have to keep my fork up to date? Or is there more that I am not realizing?

**Comment by Miffyli** [comment]
> your suggestion is that I use the unmerged forked code from this PR?

Yes, this is a suitable update upon that PR as it provides the necessary "interfaces". This means you will have to wait for that PR to be merged before you can open yours, but you can start developing already by taking that code and creating the callback.

By "merging" I mean you can make a PR about this but we won't merge it to stable-baselines3 until later versions and not before #115 is merged (which will be during v1.1).

**Comment by xicocaio** [comment]
Ok, no problem, that works for me. I will update my fork with the code from PR #115.

Also no problem with the merging happening only in the future. But I think I prefer to create the PR as soon as I am done and then keep updating it until the merge.

Thank you.

**Comment by xicocaio** [comment]
Following everything we have discussed, I implemented a solution based on PR #115 and I am now about to make a new PR. However, I should mention that I encountered a puzzling situation.

When accessing `self.locals` dict I found that, for some models, there is an entry keyword 'done' (e.g., SAC, TD3, DQN, DDPG), while some models use keyword 'dones' (e.g., A2C, PPO), which required me to check both cases as such.

```
if self.locals.get("done") or self.locals.get("dones"):
    self.current_episode += 1
```

Is there any particular reason for that difference in keywords among models?

**Comment by Miffyli** [comment]
A2C/PPO can, and probably, work on vectorized environments (`dones` is a vector of booleans), while SAC/TD3/DQN/DDPG mainly work on single environments (where `done` probably is a boolean). You might want to double-check it works for all of those algorithms. Include appropriate tests and documentation along with the callback.

**Comment by xicocaio** [comment]
Sure, I am following the contributing documentation strictly. =)

I updated the documentation and tests (which are passing). Also, I did some manual verification of behavior by inspecting variables and stdout at runtime
```
A2C reached max_episodes=1, stopping training with num_timesteps=200
PPO reached max_episodes=1, stopping training with num_timesteps=200
SAC reached max_episodes=1, stopping training with num_timesteps=200
TD3 reached max_episodes=1, stopping training with num_timesteps=200
DQN reached max_episodes=1, stopping training with num_timesteps=12
DDPG reached max_episodes=1, stopping training with num_timesteps=200
```
Thus, I verified that in both cases, the done flags keys are arrays of boolean of size 1.
```
'done': array([ True])
'dones': array([False])
```
Still, I did not test what would happen in a vectorized env such as `SubprocVecEnv` when using multiple envs. And I should probably make a test for this case. Although, I must admit I am not sure what to do in this scenario.

As I understand, training should stop for each env individually. Then, I would have to keep track of an array of `self.current_episode`. But how would I be able to identify which index in this array corresponds to the specific env that called triggered `_on_step` and that could be stopped? Am I missing something?
